proposals/xxxx-mesh-nodes.md

```diff
@@ -0,0 +1,329 @@
+shaders are supported in the same way here:
+`SV_DispatchThreadID`, `SV_GroupThreadID`, `SV_GroupIndex`, `SV_GroupID`,
+with the exception (for now) of `SV_ViewID`, which is not supported for this
+experimental feature.
+
+Outputs to other nodes are not allowed - the mesh node must be a leaf in the
+work graph.
+
+The new `mesh` node supports outputs inherited from the mesh shaders:
+`vertices`, `primitives`, and `indices`.  As with mesh shaders, the
+`SetMeshOutputCounts` intrinsic must be used before writing to these outputs.
+As with mesh shaders, the `vertices` and `primitives` outputs are combined to
+provide the inputs the pixel shader requires.
+Each of these require a user defined structure using semantics on fields, where
+fields of HLSL basic types will be used to generate signature elements in the
+same way they are for mesh shader signature handling.
+
+Amplification shaders are not supported in this context, since work graph node
+shaders provide the function of generating the mesh node dispatches on the GPU.
+
+Mesh nodes can also be used as one or more entrypoints in a work graph
+(standalone since they can't output to other nodes),
+in which case the input record (payload) is supplied from DispatchGraph()
+at the API just like any other work graph entrypoint input.
+
+A complete mesh node is defined at the API by first creating a
+[generic program](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_generic_program_desc)
+including the mesh node function name, a pixel shader, and other optional state,
+then referring to that program in a
+[work graph program node](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_program_node),
+which may optionally override node properties, such as the node ID.
+
+See the [DispatchMesh Launch Nodes](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#dispatchmesh-launch-nodes)
+section of the work graphs spec, along with sections linked from there, to
+complete the picture.
+
+In the future, we may add subobjects to specify more of these details within
+the DXIL library for convenience.
+
+### Example Syntax
+
+```cpp
+struct MyMeshRecord {
+  float3 perMeshConstant;
+  uint3 dispatchGrid : SV_DispatchGrid;
+}
+
+groupshared uint numVerts;
+groupshared uint numPrims;
+
+[shader("node")]
+[NodeLaunch("mesh")]
+[NodeIsProgramEntry]
+[NodeID("meshNode", 2)]
+[NodeLocalRootArgumentsTableIndex(4)]
+[NumThreads(4,4,1)]
+[NodeMaxDispatchGrid(16,16,1)]
+[OutputTopology("triangle")]
+void MyMeshNode(
+    DispatchNodeInputRecord<MyMeshRecord> nodeInput,
+    out vertices PerVertexValues verts[16],
+    out primitives PerPrimitiveValues prim[8],
+    out indices uint3 tris[16] )
+{
+  float3 data = nodeInput.Get().perMeshConstant;
+  // compute numVerts, numPrims...
+  SetMeshOutputCounts(numVerts, numPrims);
+  // Cooperate across group to write output arrays...
```

Comment:
I'm not sure it's clear that this implies that additional code should be here to do this. Maybe a dummy function called "writeOutputArrays()" or something.

Reply:
That's a nice idea, but writing this correctly is harder.  You do not want to pass the vertices, primitives, or indices arrays to a function to be written.  Rather, you need to do this in the main function, otherwise, you have per-thread copy-in/copy-out semantics on each array, when those arrays are essentially group shared.

I could write some code that assigns array elements values based on something like SV_GroupIndex that I add.  Like so:
```cpp
  // Cooperate across group to write output arrays...
  if (gi < numVerts) {
    verts[gi] = ComputeVertexValues(gi);
  }
  if (gi < numPrims) {
    prims[gi] = ComputePrimValues(gi);
    tris[gi] = ComputeTri(gi);
  }
```